tensorflow_probability/python/experimental/mcmc/sample_discarding_kernel_test.py

```diff
@@ -0,0 +1,294 @@
+    self.assertEqual(1, kernel_results.call_counter)
+    self.assertEqual(12, kernel_results.inner_results.counter_1)
+    self.assertEqual(24, kernel_results.inner_results.counter_2)
+
+  def test_is_calibrated(self):
+    calibrated_kernel = TestTransitionKernel()
+    uncalibrated_kernel = TestTransitionKernel(is_calibrated=False)
+    calibrated_discarder = tfp.experimental.mcmc.SampleDiscardingKernel(
+        calibrated_kernel)
+    uncalibrated_discarder = tfp.experimental.mcmc.SampleDiscardingKernel(
+        uncalibrated_kernel)
+    self.assertTrue(calibrated_discarder.is_calibrated)
+    self.assertFalse(uncalibrated_discarder.is_calibrated)
+
+  def test_with_composed_kernel(self):
+    fake_inner_kernel = TestTransitionKernel()
+    cov_reducer = tfp.experimental.mcmc.CovarianceReducer()
+    reducer_kernel = tfp.experimental.mcmc.WithReductions(
+        inner_kernel=tfp.experimental.mcmc.SampleDiscardingKernel(
+            inner_kernel=fake_inner_kernel,
+            num_burnin_steps=10,
+            num_steps_between_results=2,),
+        reducers=cov_reducer
+    )
+    current_state, kernel_results = 0., reducer_kernel.bootstrap_results(0.)
+    for _ in range(2):
+      current_state, kernel_results = reducer_kernel.one_step(
+          current_state, kernel_results)
+    current_state, kernel_results, cov = self.evaluate([
+        current_state,
+        kernel_results,
+        cov_reducer.finalize(kernel_results.streaming_calculations),
+    ])
+    self.assertEqual(16, current_state)
+    self.assertEqual(2, kernel_results.inner_results.call_counter)
+    self.assertEqual(
+        16, kernel_results.inner_results.inner_results.counter_1)
+    self.assertEqual(
+        32, kernel_results.inner_results.inner_results.counter_2)
+    self.assertEqual(np.var([13, 16]), cov)
+
+  def test_tf_while(self):
+    fake_inner_kernel = TestTransitionKernel()
+    discarder = tfp.experimental.mcmc.SampleDiscardingKernel(
+        fake_inner_kernel,
+        num_burnin_steps=10,
+        num_steps_between_results=1,)
+
+    def _loop_body(i, curr_state, pkr):
+      new_state, kernel_results = discarder.one_step(
+          curr_state, pkr,
+      )
+      return (i + 1, new_state, kernel_results)
+
+    pkr = discarder.bootstrap_results(0.)
+    _, final_sample, kernel_results = tf.while_loop(
+        lambda i, *_: i < 2,
+        _loop_body,
+        (0., 0., pkr),
+    )
+    final_sample, kernel_results = self.evaluate([
+        final_sample, kernel_results])
+    self.assertEqual(14, final_sample)
+    self.assertEqual(2, kernel_results.call_counter)
+    self.assertEqual(14, kernel_results.inner_results.counter_1)
+    self.assertEqual(28, kernel_results.inner_results.counter_2)
+
+  def test_tensor_thinning_and_burnin(self):
+    fake_inner_kernel = TestTransitionKernel()
+    discarder = tfp.experimental.mcmc.SampleDiscardingKernel(
+        fake_inner_kernel,
+        num_burnin_steps=tf.convert_to_tensor(10),
```

Comment:
If this was meant to test the non-statically known pathway, it doesn't as these are trivially statically known. I recommend replacing them with tf.Variable.